### PR TITLE
introduce closure to mutate resource or file id for dom use only

### PR DESCRIPTION
**Pull request by fschade**
## Description
in some cases the resource or item id is not a valid dom selector, to be able to mutate the id this introduces a closure to do so.

## Related Issue

- https://github.com/owncloud/web/issues/5927

## Motivation and Context
valid dom selectors

## How Has This Been Tested?

- unit tests
- yarn link and web

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Tests

## Checklist:

- [X] Code changes
- [X] Unit tests added
- [X] Documentation added/updated

## Open tasks:

- [ ] confirmation if it fixes https://github.com/owncloud/web/issues/5927

### CODE REVIEW

**Review by pascalwengerter — CHANGES_REQUESTED** [review]
I added an offending ID in the examples (e.g. line 641 `id: "forest:19",`) to test this, and it visibly breaks the component (by always rendering the dropdown menu) even if the regex itself does the job - string "aa:a/a-a" gets turned into "aaaaa"

**Comment by pascalwengerter** [comment]
> I added an offending ID in the examples (e.g. line 641 `id: "forest:19",`) to test this, and it visibly breaks the component (by always rendering the dropdown menu) even if the regex itself does the job - string "aa:a/a-a" gets turned into "aaaaa"

Would be great to have this covered in test data somehow, too 😇

File: src/components/molecules/OcTable/OcTable.spec.js [hunk 1]
```diff
@@ -195,15 +195,30 @@ describe("OcTable", () => {
...
+  it("adds data-item-id for rows", () => {
     const wrapper = shallowMount(Table, {
       propsData: {
         fields,
         data,
         highlighted: [],
       },
     })
-    expect(wrapper.html().indexOf("data-file-id")).toBeGreaterThan(-1)
+    console.log(wrapper.html())
```
Comment (pascalwengerter) [inline src/components/molecules/OcTable/OcTable.spec.js:206]: Dev leftover? 🐶

**Comment by fschade** [comment]
> I added an offending ID in the examples (e.g. line 641 `id: "forest:19",`) to test this, and it visibly breaks the component (by always rendering the dropdown menu) even if the regex itself does the job - string "aa:a/a-a" gets turned into "aaaaa"

nice thanks, yes this should fail in line 641, the example overwrites the `resourceDomSelector` [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.vue#L705) with an unsafe implementation, if you remove the overwrite will fallback [to](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.vue#L169) and will work.

the case is covered [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.spec.js#L161) for an overwrite and [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.spec.js#L229) for the default behavior. Thanks ❤️  for deep dive

**Comment by pascalwengerter** [comment]
> > I added an offending ID in the examples (e.g. line 641 `id: "forest:19",`) to test this, and it visibly breaks the component (by always rendering the dropdown menu) even if the regex itself does the job - string "aa:a/a-a" gets turned into "aaaaa"
> 
> nice thanks, yes this should fail in line 641, the example overwrites the `resourceDomSelector` [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.vue#L705) with an unsafe implementation, if you remove the overwrite it will fallback to [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.vue#L169) and will work.
> 
> the case is covered [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.spec.js#L161) for an overwrite and [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.spec.js#L229) for the default behavior. Thanks for your deep dive heart

Okay now I've read the actual implementation 🤦🏽 if CERN confirms this works with their EOS IDs (which it should) I'm happy to approve, merge & release!

**Comment by pascalwengerter** [comment]
> > > I added an offending ID in the examples (e.g. line 641 `id: "forest:19",`) to test this, and it visibly breaks the component (by always rendering the dropdown menu) even if the regex itself does the job - string "aa:a/a-a" gets turned into "aaaaa"
> > 
> > 
> > nice thanks, yes this should fail in line 641, the example overwrites the `resourceDomSelector` [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.vue#L705) with an unsafe implementation, if you remove the overwrite it will fallback to [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.vue#L169) and will work.
> > the case is covered [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.spec.js#L161) for an overwrite and [here](https://github.com/owncloud/owncloud-design-system/blob/3d31231dff597c0bce0db4cb3e7876270ac51d4a/src/components/templates/OcTableFiles/OcTableFiles.spec.js#L229) for the default behavior. Thanks for your deep dive heart
> 
> Okay now I've read the actual implementation 🤦🏽 if CERN confirms this works with their EOS IDs (which it should) I'm happy to approve, merge & release!

Not sure how smart it is to overwrite the sanitization function in an example though, creates knots in the heads of people like me haha

File: src/components/molecules/OcTable/OcTable.spec.js (hunk 1 above)
Comment (kulmann) [reply src/components/molecules/OcTable/OcTable.spec.js:206]: Same question from me 😁 Otherwise looks good..

Comment (fschade) [reply src/components/molecules/OcTable/OcTable.spec.js]: as usual, get things out fast to test ci and forget to remove those leftovers :( sorry guys, cleaned

**Comment by diocas** [comment]
@fschade just tested and it fixes the issue. (the `0-9\-_` that you added after was important :))
Not sure if hashing would be more future proof, but is reasonable to expect that letter and numbers would suffice.

**Comment by fschade** [comment]
@kulmann, @pascalwengerter please have a look, i also touched the naming if `..file..` to `..item..` in this PR.
I know that it's not related to this PR's topic but popped up in the rebase

**Review by pascalwengerter — APPROVED** [review]
Changes LGTM, thanks for the additional cleanup!

File: src/components/templates/OcTableFiles/OcTableFiles.vue [hunk 2]
```diff
@@ -13,8 +13,8 @@
     @highlight="fileClicked"
     @rowMounted="rowMounted"
     @contextmenuClicked="showContextMenu"
-    @fileDropped="fileDropped"
-    @fileDragged="fileDragged"
+    @itemDropped="fileDropped"
```
Comment (pascalwengerter) [inline src/components/templates/OcTableFiles/OcTableFiles.vue:16]: This was an unnecessary breaking change in retrospect (it's the OcFilesTable so file as a prefix is fine, slipped through my code review), let's discuss if we want to rollback & re-release or ship a bugfix release tomorrow

**Pull request merged**